File: ReactNative/src/services/transactionService.ts
import { Account, Transaction, TransactionClassification, BudgetCategory } from '../models/types';
import { PlaidTransaction, fetchTransactions } from './plaidService';
import { markAccountSynced, removeTransactionsByPlaidId, upsertTransactions } from '../database';

/**
 * ANTI-DOUBLE-COUNT CLASSIFICATION
 * 
 * The #1 problem with tracking spending across multiple credit cards:
 * 
 *   You buy $50 at Target on Chase card     → SPENDING (counts toward budget)
 *   You pay Chase $500 from checking        → TRANSFER (excluded from budget)
 *   You transfer $583 to Vanguard Roth IRA  → CONTRIBUTION (tracked separately)
 *   Your paycheck hits checking             → INCOME (cash flow only)
 * 
 * Without this, the $50 Target purchase gets counted TWICE:
 * once as the CC charge, once as part of the $500 CC payment.
 */
export function classifyTransaction(
  plaidTx: PlaidTransaction,
  account: Account
): TransactionClassification {
  const name = (plaidTx.merchant_name ?? plaidTx.name).toLowerCase();
  const primary = plaidTx.personal_finance_category?.primary ?? '';
  const detailed = plaidTx.personal_finance_category?.detailed ?? '';

  // ── INCOME ──
  if (plaidTx.amount < 0 && account.accountType === 'depository') {
    if (isInternalTransfer(primary, detailed, name)) return 'transfer';
    return 'income';
  }

  // ── CREDIT CARD PAYMENTS ──
  if (isCreditCardPayment(primary, detailed, name)) return 'transfer';

  // ── INTERNAL TRANSFERS ──
  if (isInternalTransfer(primary, detailed, name)) {
    if (isInvestmentTransfer(name, account)) return 'contribution';
    return 'transfer';
  }

  // ── RETIREMENT/INVESTMENT ACCOUNTS ──
  if (account.isRetirementAccount) return 'contribution';
  if (isInvestmentTransfer(name, account)) return 'contribution';

  // ── CC PAYMENT RECEIVED (on the CC side) ──
  if (account.accountType === 'credit' && plaidTx.amount < 0) {
    if (isPaymentReceived(name)) return 'transfer';
  }

  // ── SPENDING ──
  if (plaidTx.amount > 0) return 'spending';

  return 'excluded';
}

// ── Detection Helpers ──

const CC_PAYMENT_KEYWORDS = [
  'payment', 'autopay', 'bill pay', 'credit card payment',
  'chase', 'amex', 'discover', 'citi', 'capital one',
  'barclays', 'wells fargo card', 'bank of america card',
];

function isCreditCardPayment(primary: string, detailed: string, name: string): boolean {
  if (primary === 'LOAN_PAYMENTS' || detailed.includes('CREDIT_CARD')) return true;
  const isPayment = name.includes('payment') || name.includes('autopay') || name.includes('bill pay');
  const isToCC = CC_PAYMENT_KEYWORDS.some(kw => name.includes(kw));
  return isPayment && isToCC;
}

function isInternalTransfer(primary: string, detailed: string, name: string): boolean {
  if (['TRANSFER_IN', 'TRANSFER_OUT'].includes(primary)) return true;
  const hasTransferWord = name.includes('transfer') || name.includes('xfer');
  return hasTransferWord && (primary.includes('TRANSFER') || name.includes('transfer'));
}

const INVESTMENT_KEYWORDS = [
  'vanguard', 'fidelity', 'schwab', 'tsp', 'thrift savings',
  '401k', '401(k)', 'roth', 'ira', 'brokerage',
  'etrade', 'e*trade', 'robinhood', 'wealthfront', 'betterment',
  'contribution', 'retirement', 'investment',
];

function isInvestmentTransfer(name: string, account: Account): boolean {
  if (account.accountType === 'retirement' || account.accountType === 'investment') return true;
  return INVESTMENT_KEYWORDS.some(kw => name.includes(kw));
}

function isPaymentReceived(name: string): boolean {
  const keywords = ['payment', 'thank you', 'autopay', 'online payment', 'mobile payment'];
  return keywords.some(kw => name.includes(kw));
}

// ── Category Mapping ──

export function mapToCategory(
  plaidTx: PlaidTransaction,
  classification: TransactionClassification,
  categories: BudgetCategory[]
): string {
  // Non-spending gets a fixed label
  if (classification === 'transfer') return 'Transfer';
  if (classification === 'contribution') return 'Investment';
  if (classification === 'income') return 'Income';
  if (classification === 'excluded') return 'Other';

  // Try Plaid category
  const plaidDetailed = plaidTx.personal_finance_category?.detailed;
  if (plaidDetailed) {
    const match = categories.find(c => c.plaidCategories.includes(plaidDetailed));
    if (match) return match.name;
  }

  // Keyword matching
  const searchText = (plaidTx.merchant_name ?? plaidTx.name).toLowerCase();
  for (const cat of categories) {
    if (cat.matchKeywords.some(kw => searchText.includes(kw.toLowerCase()))) {
      return cat.name;
    }
  }

  return 'Misc';
}

// ── Build Transaction Object ──

export function buildTransaction(
  plaidTx: PlaidTransaction,
  account: Account,
  categories: BudgetCategory[]
): Transaction {
  const classification = classifyTransaction(plaidTx, account);
  const categoryName = mapToCategory(plaidTx, classification, categories);

  return {
    id: plaidTx.transaction_id,
    plaidTransactionId: plaidTx.transaction_id,
    date: plaidTx.date,
    amount: plaidTx.amount,
    merchantName: plaidTx.merchant_name,
    name: plaidTx.name,
    categoryName,
    plaidCategory: plaidTx.personal_finance_category?.detailed ?? null,
    isPending: plaidTx.pending,
    classification,
    isManuallyClassified: false,
    notes: null,
    accountId: account.id,
  };
}

// ── Sync (fetch from Plaid, classify, persist) ──
//
// Mirrors the anti-double-count sync loop in the iOS TransactionService.

export async function syncAccountTransactions(
  account: Account,
  categories: BudgetCategory[]
): Promise<void> {
  let cursor: string | undefined;
  let hasMore = true;

  while (hasMore) {
    const response = await fetchTransactions(account.plaidAccessToken, cursor);

    const newTransactions = response.added.map(plaidTx =>
      buildTransaction(plaidTx, account, categories)
    );
    await upsertTransactions(newTransactions);

    if (response.removed.length > 0) {
      await removeTransactionsByPlaidId(response.removed.map(r => r.transaction_id));
    }

    cursor = response.next_cursor;
    hasMore = response.has_more;
  }

  await markAccountSynced(account.id);
}

File: PinkBudget/Services/TransactionService.swift
import Foundation
import SwiftData

/// Handles syncing Plaid transactions into SwiftData,
/// auto-categorizing, detecting contributions, and
/// preventing double-counting across credit cards and transfers.
///
/// CLASSIFICATION RULES:
/// ─────────────────────
/// 1. SPENDING: Individual purchases on credit/debit cards.
///    These are the ONLY transactions that count toward your budget.
///
/// 2. TRANSFER: CC payments, account-to-account moves.
///    EXCLUDED from budget to prevent double-counting.
///    Example: You pay $500 from checking to Chase CC.
///    The $500 is NOT spending -- the individual CC charges are.
///
/// 3. CONTRIBUTION: Money going to investment/retirement accounts.
///    Tracked in the Contributions tab, NOT as spending.
///    Example: $583/mo to Vanguard Roth IRA.
///
/// 4. INCOME: Paychecks, deposits.
///    Tracked for cash flow, not counted as spending.
///
/// HOW DOUBLE-COUNTING IS PREVENTED:
/// ──────────────────────────────────
/// - CC payments from checking → classified as TRANSFER → excluded
/// - The actual CC purchases (Target $50, Chipotle $12) → SPENDING → budgeted
/// - Vanguard/TSP/401k transfers → CONTRIBUTION → tracked separately
/// - Internal transfers (savings ↔ checking) → TRANSFER → excluded

class TransactionService {
    private let modelContext: ModelContext
    
    init(modelContext: ModelContext) {
        self.modelContext = modelContext
    }
    
    // MARK: - Transaction Classification
    
    enum TransactionClass {
        case spending       // Counts toward budget
        case transfer       // Excluded (CC payments, internal moves)
        case contribution   // Retirement/investment deposits
        case income         // Paychecks, refunds
        case excluded       // Fees, interest, other non-budget items
    }
    
    /// Classify a Plaid transaction to prevent double-counting
    func classify(_ plaidTx: PlaidTransaction, account: Account) -> TransactionClass {
        let name = (plaidTx.merchantName ?? plaidTx.name).lowercased()
        let plaidPrimary = plaidTx.personalFinanceCategory?.primary ?? ""
        let plaidDetailed = plaidTx.personalFinanceCategory?.detailed ?? ""
        
        // ── INCOME ──
        // Negative amounts on depository accounts = money coming in
        if plaidTx.amount < 0 && account.accountType == .depository {
            // But not if it's a transfer from another account
            if isInternalTransfer(plaidPrimary, plaidDetailed, name) {
                return .transfer
            }
            // Check if it's a refund on a debit card
            if isRefund(plaidPrimary, name) {
                return .income  // or .excluded, depending on preference
            }
            return .income
        }
        
        // ── CREDIT CARD PAYMENTS ──
        // Payment FROM checking/savings TO a credit card = transfer
        if isCreditCardPayment(plaidPrimary, plaidDetailed, name) {
            return .transfer
        }
        
        // ── INTERNAL TRANSFERS ──
        // Moving money between your own accounts
        if isInternalTransfer(plaidPrimary, plaidDetailed, name) {
            // Special case: transfers TO investment/retirement accounts = contributions
            if isInvestmentTransfer(name, account) {
                return .contribution
            }
            return .transfer
        }
        
        // ── CONTRIBUTIONS ──
        // Money going to retirement/investment accounts
        if account.isRetirementAccount {
            if plaidTx.amount > 0 {
                // Debit from checking TO retirement = contribution
                return .contribution
            } else {
                // Credit on the retirement account side
                return .contribution
            }
        }
        if isInvestmentTransfer(name, account) {
            return .contribution
        }
        
        // ── CREDIT CARD: PAYMENT RECEIVED ──
        // On the CC side, a payment shows as negative (credit)
        if account.accountType == .credit && plaidTx.amount < 0 {
            if isPaymentReceived(name) {
                return .transfer  // Don't count CC payment as income
            }
        }
        
        // ── SPENDING ──
        // Everything else that's a positive amount (debit) = real spending
        if plaidTx.amount > 0 {
            return .spending
        }
        
        return .excluded
    }
    
    // MARK: - Detection Helpers
    
    private func isCreditCardPayment(_ primary: String, _ detailed: String, _ name: String) -> Bool {
        let ccPaymentKeywords = [
            "payment", "autopay", "bill pay", "credit card payment",
            "chase", "amex", "discover", "citi", "capital one",
            "barclays", "wells fargo card", "bank of america card"
        ]
        
        // Plaid explicitly categorizes these
        if primary == "LOAN_PAYMENTS" || detailed.contains("CREDIT_CARD") {
            return true
        }
        
        // Keyword fallback
        let isPayment = name.contains("payment") || name.contains("autopay") || name.contains("bill pay")
        let isToCC = ccPaymentKeywords.contains { name.contains($0) }
        
        return isPayment && isToCC
    }
    
    private func isInternalTransfer(_ primary: String, _ detailed: String, _ name: String) -> Bool {
        // Plaid categories for transfers
        let transferCategories = ["TRANSFER_IN", "TRANSFER_OUT"]
        if transferCategories.contains(primary) {
            return true
        }
        
        let transferKeywords = [
            "transfer", "xfer", "ach", "wire",
            "zelle", "venmo", "paypal",  // These may or may not be transfers
            "online banking transfer", "mobile transfer",
            "savings transfer", "checking transfer"
        ]
        
        // Only flag as transfer if it looks like an internal move
        // (not a Venmo payment for dinner, which IS spending)
        return transferKeywords.contains { name.contains($0) } &&
               (name.contains("transfer") || name.contains("xfer") || primary.contains("TRANSFER"))
    }
    
    private func isInvestmentTransfer(_ name: String, _ account: Account) -> Bool {
        let investmentKeywords = [
            "vanguard", "fidelity", "schwab", "tsp", "thrift savings",
            "401k", "401(k)", "roth", "ira", "brokerage",
            "etrade", "e*trade", "robinhood", "wealthfront", "betterment",
            "contribution", "retirement", "investment"
        ]
        
        // If the account is explicitly a retirement/investment account
        if account.accountType == .retirement || account.accountType == .investment {
            return true
        }
        
        return investmentKeywords.contains { name.contains($0) }
    }
    
    private func isRefund(_ primary: String, _ name: String) -> Bool {
        return primary == "REFUND" || name.contains("refund") || name.contains("return")
    }
    
    private func isPaymentReceived(_ name: String) -> Bool {
        let paymentKeywords = [
            "payment", "thank you", "autopay",
            "online payment", "mobile payment", "ach payment"
        ]
        return paymentKeywords.contains { name.contains($0) }
    }
    
    // MARK: - Sync Transactions
    
    func syncTransactions(for account: Account) async throws {
        let plaid = PlaidService.shared
        var cursor: String? = nil
        var hasMore = true
        
        while hasMore {
            let response = try await plaid.fetchTransactions(
                accessToken: account.plaidAccessToken,
                cursor: cursor
            )
            
            for plaidTx in response.added {
                // Skip duplicates
                let existing = try modelContext.fetch(
                    FetchDescriptor<Transaction>(
                        predicate: #Predicate { $0.plaidTransactionId == plaidTx.transactionId }
                    )
                )
                guard existing.isEmpty else { continue }
                
                // Classify the transaction
                let classification = classify(plaidTx, account: account)
                
                let transaction = Transaction(
                    plaidTransactionId: plaidTx.transactionId,
                    date: parseDate(plaidTx.date),
                    amount: plaidTx.amount,
                    merchantName: plaidTx.merchantName,
                    name: plaidTx.name,
                    categoryName: mapCategory(plaidTx, classification: classification),
                    plaidCategory: plaidTx.personalFinanceCategory?.detailed,
                    isPending: plaidTx.pending,
                    isContribution: classification == .contribution,
                    isExcludedFromBudget: classification != .spending
                )
                transaction.account = account
                modelContext.insert(transaction)
                
                // Auto-create Contribution record
                if classification == .contribution {
                    let contribution = Contribution(
                        date: transaction.date,
                        amount: abs(transaction.amount),
                        label: account.contributionLabel,
                        source: .plaid
                    )
                    contribution.linkedTransaction = transaction
                    modelContext.insert(contribution)
                }
            }
            
            // Handle removed transactions
            for removed in response.removed {
                let existing = try modelContext.fetch(
                    FetchDescriptor<Transaction>(
                        predicate: #Predicate { $0.plaidTransactionId == removed.transactionId }
                    )
                )
                for tx in existing {
                    modelContext.delete(tx)
                }
            }
            
            cursor = response.nextCursor
            hasMore = response.hasMore
        }
        
        account.lastSynced = Date()
        try modelContext.save()
    }
    
    // MARK: - Category Mapping
    
    private func mapCategory(_ plaidTx: PlaidTransaction, classification: TransactionClass) -> String {
        // Non-spending gets a special category label
        switch classification {
        case .transfer:
            return "Transfer"
        case .contribution:
            return "Investment"
        case .income:
            return "Income"
        case .excluded:
            return "Excluded"
        case .spending:
            break  // Fall through to normal categorization
        }
        
        // Try Plaid category mapping
        if let plaidCat = plaidTx.personalFinanceCategory?.detailed {
            if let categories = try? modelContext.fetch(FetchDescriptor<BudgetCategory>()) {
                for cat in categories {
                    if cat.plaidCategories.contains(plaidCat) {
                        return cat.name
                    }
                }
            }
        }
        
        // Keyword matching
        let searchText = (plaidTx.merchantName ?? plaidTx.name).lowercased()
        if let categories = try? modelContext.fetch(FetchDescriptor<BudgetCategory>()) {
            for cat in categories {
                for keyword in cat.matchKeywords {
                    if searchText.contains(keyword.lowercased()) {
                        return cat.name
                    }
                }
            }
        }
        
        return "Misc"
    }
    
    private func parseDate(_ dateString: String) -> Date {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.date(from: dateString) ?? Date()
    }
}
